src/lyra/abstract_domains/string/character_domain.py

```diff
@@ -0,0 +1,263 @@
+import string
+from copy import deepcopy
+
+from docutils.io import Input
+
+from lyra.abstract_domains.lattice import BottomMixin
+from lyra.abstract_domains.state import State
+from lyra.abstract_domains.store import Store
+from lyra.core.expressions import *
+from lyra.core.types import StringLyraType
+from lyra.core.utils import copy_docstring
+
+_alphabet = set(string.printable)
+
+
+class CharacterLattice(BottomMixin):
+
+    def __init__(self, certainly=set(), maybe=_alphabet):
+        self._certainly = certainly
+        self._maybe = maybe
+        super().__init__()
+        if not self.certainly.issubset(self.maybe):
+            self.bottom()
+
+    @property
+    def certainly(self):
+        return self._certainly
+
+    @property
+    def maybe(self):
+        return self._maybe
+
+    def top(self):
+        return self._replace(CharacterLattice())
+
+    def __repr__(self):
+        if self.is_bottom():
+            return "⊥"
+
+        if self.is_top():
+            return "T"
+
+        return "(<{}>, <{}>)".format(self.certainly, self.maybe)
+
+    @copy_docstring(BottomMixin.is_top)
+    def is_top(self) -> bool:
+        return not self.is_bottom() and self.certainly == set() and self.maybe == _alphabet
```

Comment:
I'm not sure about this because the ``bottom()`` function in ``BottomMixin`` changes the ``_kind`` attribute without changing any other attributes. The statement ``entry.bottom()`` in the analysis is called on a state for which ``maybe`` and ``certainly are already initialized. In this case, ``is_top()`` will evaluate to true. I remember I had such a bug before.